### PR TITLE
Sub: Allow attitude control via mavlink in althold

### PR DESCRIPTION
**Pull request by patrickelectric**
- Update some ArduSub code style to match rest of the project (millis, micros)
- Allow arbitrary attitude control from SET_ATTITUDE_TARGET without a gps system

### CODE REVIEW

File: ArduSub/control_althold.cpp [hunk 1]
```diff
@@ -49,7 +49,30 @@ void Sub::althold_run()
 
     // get pilot desired lean angles
     float target_roll, target_pitch;
-    get_pilot_desired_lean_angles(channel_roll->get_control_in(), channel_pitch->get_control_in(), target_roll, target_pitch, attitude_control.get_althold_lean_angle_max());
+
+    // Check if set_attitude_target_no_gps is valid
+    if (tnow - sub.set_attitude_target_no_gps.status.last_message_ms < 5e3 && sub.set_attitude_target_no_gps.status.is_valid) {
```
Comment (OXINARF) [inline ArduSub/control_althold.cpp]: Using 5e3 makes it a float comparison unnecessarily.

**Comment by patrickelectric** [comment]
Ty for pointing this @OXINARF !

File: ArduSub/Sub.h [hunk 2]
```diff
@@ -439,6 +439,20 @@ class Sub : public AP_HAL::HAL::Callbacks {
...
+        uint32_t last_message_ms;
+    };
+
+    struct attitude_no_gps_struct {
+        command_validation_struct status;
+        mavlink_set_attitude_target_t packet;
+    };
+
+    attitude_no_gps_struct set_attitude_target_no_gps {{false, 0}};
+    command_validation_struct set_position_target_local_ned_no_gps {false, 0};
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply ArduSub/Sub.h]: Done.

File: ArduSub/control_althold.cpp [hunk 3]
```diff
@@ -49,7 +49,30 @@ void Sub::althold_run()
...
+        ).to_euler(
+            target_roll,
+            target_pitch,
+            target_yaw
+        );
+        target_roll = degrees(target_roll);
+        target_pitch = degrees(target_pitch);
+        target_yaw = degrees(target_yaw);
+
+        attitude_control.input_euler_angle_roll_pitch_yaw(target_roll * 1e2, target_pitch * 1e2, target_yaw * 1e2, true);
```
Comment (jaxxzer) [inline ArduSub/control_althold.cpp]: put float specifier on constants

File: ArduSub/control_althold.cpp [hunk 4]
```diff
@@ -49,7 +49,30 @@ void Sub::althold_run()
...
+        return;
+
+    } else {
+        // Set valid to false, avoiding overflow in last_message_ms
+        sub.set_attitude_target_no_gps.status.is_valid = false;
```
Comment (jaxxzer) [inline ArduSub/control_althold.cpp]: I do not see why it is necessary to have the is_valid flag. We can just check the timestamp.

**Comment by jaxxzer** [comment]
We can add it to stabilize mode too if you want @patrickelectric.

**Pull request merged**